Remove restriction on anonymous read-only volumes

srstsavage opened the pull request:
Remove restriction on anonymous read-only volumes.

This restriction is currently preventing the use of pre-populated volumes that should be accessed in a read-only manner in a container (e.g. an NFS volume containing data to be processed or served).

According to @neersighted the restriction may have originally been put in place with the assumption that pre-populated volumes would be exposed as a named volume by the volume driver.

In practice, NFS volumes are mounted using the docker `local` driver by supplying driver opts. Example that fails when `readonly` is specified but works without:

```
docker run --rm -it \
 --mount 'readonly,type=volume,dst=/data/dest,volume-driver=local,volume-opt=type=nfs,volume-opt=device=:/export/some-share,"volume-opt=o=nfsvers=4,addr=some.server"' \
  debian
```

Fixes #45297

**- What I did**
Removed restriction on anonymous read-only volumes.

**- How I did it**
Deleted simple checks for anonymous read-only volumes from Linux and Windows volume parsers.

**- How to verify it**
After changes, read only mount an NFS share using something like

```
docker run --rm -it \
 --mount 'readonly,type=volume,dst=/data/dest,volume-driver=local,volume-opt=type=nfs,volume-opt=device=:/export/some-share,"volume-opt=o=nfsvers=4,addr=some.server"' \
  debian
```

**- Human readable description for the release notes**
```markdown changelog
Remove restriction on anonymous read-only volumes
```

<img width="494" height="461" alt="image" src="https://github.com/user-attachments/assets/bf70cf50-e25a-4631-a641-f034d00d7f8e" />

srstsavage committed:
Remove restriction on anonymous read-only volumes

Restriction on anonymouse read-only volumes is currently preventing
the use of pre-populated volumes that should be accessed in a read-only manner in a container
(e.g. an NFS volume containing data to be processed or served).

According to @neersighted the restriction may have originally been put
in place with the assumption that pre-populated volumes would be
exposed as a named volume by the volume driver.

In practice, NFS volumes are mounted using the docker `local` driver
by supplying driver opts. Example that fails when `readonly` is specified but works without:

```
docker run --rm -it \
 --mount 'readonly,type=volume,dst=/data/dest,volume-driver=local,volume-opt=type=nfs,volume-opt=device=:/export/some-share,"volume-opt=o=nfsvers=4,addr=some.server"' \
  debian
```

Fixes #45297

Signed-off-by: Shane St Savage <shane@axds.co>

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
We should add some tests for this though before merging; let me know if you need help with that/

thaJeztah commented:
I gave CI a kick to run

thaJeztah commented on daemon/volume/mounts/linux_parser_test.go:
```diff
@@ -264,6 +264,80 @@ func TestLinuxParseMountRawSplit(t *testing.T) {
+	}{
+		{
+			mount: mount.Mount{
+				Source: "/tmp",
```
Then use that here;


```suggestion
				Source: tmpDir,
```

vvoland committed:
daemon/volume/mounts: Fix tests on Windows

The test was failing on Windows because it used Unix-style paths and
relied on platform-specific filesystem behavior.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
I've adjusted the tests (one is skipped on Windows).

It's a bit of a workaround, because the parser itself shouldn't care if the source directory exists or not... but it's unrelated to this change so let's not try to fix it here.

vvoland reviewed (approved):
LGTM, but since I pushed one commit, I would appreciate a second review 😅

thaJeztah reviewed (approved):
LGTM
